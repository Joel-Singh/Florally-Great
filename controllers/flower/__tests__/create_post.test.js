const { validationResult } = require("express-validator");
const emulateCallingController = require("../../testingUtils/emulateCallingController");
const create_post = require("../create_post");

describe("Test validation", () => {
  const flowerProperties = ["name", "description", "numberInStock", "price"];
  test.each(flowerProperties)(
    `Doesn't accept empty input for %s`,
    async (property) => {
      expect(
        await getFilteredValidationErrors(property, "", "empty")
      ).toMatchSnapshot();
    }
  );

  test.each(flowerProperties)(
    `Doesn't accept just spaces for %s`,
    async (property) => {
      expect(
        await getFilteredValidationErrors(property, "   ", "empty")
      ).toMatchSnapshot();
    }
  );

  test(`Only accepts numbers for number in stock`, async () => {
    expect(
      await getFilteredValidationErrors("numberInStock", 32, "number")
    ).toEqual([]);

    expect(
      await getFilteredValidationErrors(
        "numberInStock",
        "not a number",
        "number"
      )
    ).toMatchSnapshot();
  });

  test(`Only accepts property formatted prices`, async () => {
    expect(
      await getFilteredValidationErrors("price", 32, "format")
    ).toMatchSnapshot();

    expect(
      await getFilteredValidationErrors("price", "$3.86", "format")
    ).toEqual([]);
  });
});

async function getFilteredValidationErrors(name, value, msgMustInclude) {
  const { fakeReq } = await emulateCallingController(create_post, {
    body: {
      [name]: value,
    },
  });

  const resultOfValidation = validationResult(fakeReq);
  const errorsForJustProperty = resultOfValidation.errors.filter(
    ({ path, msg }) => {
      return path === name && msg.includes(msgMustInclude);
    }
  );

  return errorsForJustProperty;
}
