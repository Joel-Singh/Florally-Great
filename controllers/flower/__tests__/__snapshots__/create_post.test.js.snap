// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test validation Doesn't accept empty input for description 1`] = `
[
  {
    "location": "body",
    "msg": "Description can't be empty",
    "path": "description",
    "type": "field",
    "value": "",
  },
]
`;

exports[`Test validation Doesn't accept empty input for name 1`] = `
[
  {
    "location": "body",
    "msg": "Name can't be empty",
    "path": "name",
    "type": "field",
    "value": "",
  },
]
`;

exports[`Test validation Doesn't accept empty input for numberInStock 1`] = `
[
  {
    "location": "body",
    "msg": "Number in stock can't be empty",
    "path": "numberInStock",
    "type": "field",
    "value": "",
  },
]
`;

exports[`Test validation Doesn't accept empty input for price 1`] = `
[
  {
    "location": "body",
    "msg": "Price can't be empty",
    "path": "price",
    "type": "field",
    "value": "",
  },
]
`;

exports[`Test validation Doesn't accept just spaces for description 1`] = `
[
  {
    "location": "body",
    "msg": "Description can't be empty",
    "path": "description",
    "type": "field",
    "value": "",
  },
]
`;

exports[`Test validation Doesn't accept just spaces for name 1`] = `
[
  {
    "location": "body",
    "msg": "Name can't be empty",
    "path": "name",
    "type": "field",
    "value": "",
  },
]
`;

exports[`Test validation Doesn't accept just spaces for numberInStock 1`] = `
[
  {
    "location": "body",
    "msg": "Number in stock can't be empty",
    "path": "numberInStock",
    "type": "field",
    "value": "",
  },
]
`;

exports[`Test validation Doesn't accept just spaces for price 1`] = `
[
  {
    "location": "body",
    "msg": "Price can't be empty",
    "path": "price",
    "type": "field",
    "value": "",
  },
]
`;

exports[`Test validation Only accepts numbers for number in stock 1`] = `
[
  {
    "location": "body",
    "msg": "Number in stock needs to be a number",
    "path": "numberInStock",
    "type": "field",
    "value": "not a number",
  },
]
`;

exports[`Test validation Only accepts property formatted prices 1`] = `
[
  {
    "location": "body",
    "msg": "Price needs to be in $x.xx format, e.g $3.86 or $287.00",
    "path": "price",
    "type": "field",
    "value": "32",
  },
]
`;
